File: processor/addressing.py
import hashlib

def _hash(string):
    return hashlib.sha512(string.encode('utf-8')).hexdigest()

def _make_touchpoint_postfix(num):
    return hex(num)[2:].zfill(4)

# The first six characters of a History or Asset address are
# the first six characters of the transaction family ("lace")
# hashed.  The next character indicates if the object is a 
# history or asset.  ~~touchpoints?

FAMILY_NAME = 'lace'

NAMESPACE = _hash(FAMILY_NAME)[:6]

ASSET = '0'
HISTORY = '1'
AGENT = '2'

def make_asset_address(identifier):
    return (
        NAMESPACE
        + ASSET
        + _hash(identifier)[:59]    # what is identifier? rename this
        + '0000'
    )

def make_history_address(identifier):
    return (
        NAMESPACE
        + HISTORY
        + _hash(identifier)[:59]    # TF:6 + ID:1 + hash:59 + log:4
        + '0000'
    )

def make_touchpoint_address(identifier, index):
    return (
        NAMESPACE
        + HISTORY
        + _hash(identifier)[:59]    # TF:6 + ID:1 + hash:59 + log:4
        + _make_touchpoint_postfix(index) # index of the wanted touchpoint
    )

def make_agent_address(identifier): 
    return (
        NAMESPACE
        + AGENT
        + _hash(identifier)[:63]    # TF:6 + ID:1 + hash: 62
    )
    

File: addressing.js
var crypto = require('crypto');

const FAMILY_NAME = 'lace';
const NAMESPACE = hash(FAMILY_NAME).slice(0, 6);
const ASSET = '0';
const HISTORY = '1';
const AGENT = '2';

function hash(string) {
    return crypto.createHash('sha512').update(string).digest("hex");
}

function makeTouchpointPostfix(num) {
    return num.toString(16).padStart(4, "0");
}

function makeAssetAddress(identifier) {
    return  NAMESPACE + 
            ASSET + 
            hash(identifier).slice(0, 59) + 
            '0000';
}

function makeHistoryAddress(identifier) {
    return  NAMESPACE + 
            HISTORY + 
            hash(identifier).slice(0, 59) + 
            '0000';
}

function makeTouchpointAddress(identifier, index) {
    return  NAMESPACE + 
            HISTORY + 
            hash(identifier).slice(0, 59) + 
            makeTouchpointPostfix(index);
}

function makeAgentAddress(identifier) {
    return  NAMESPACE + 
            AGENT + 
            hash(identifier).slice(0, 63);
}

module.exports={
    makeAgentAddress,
    makeAssetAddress,
    makeHistoryAddress,
    makeTouchpointAddress
}